bioimageio/core/resource_io/utils.py

```diff
@@ -54,6 +57,68 @@ def generic_visit(self, node):
             super().generic_visit(node)
 
 
+def get_sha256(path: os.PathLike) -> str:
+    h = hashlib.sha256()
+    with open(path, "rb") as f:
+        while True:
+            block = f.read(h.block_size)
+            if not block:
+                break
+            h.update(block)
+
+    return h.hexdigest()
+
+
+class Sha256NodeChecker(NodeVisitor):
+    """Check integrity of the source-like field for every sha256-like field encountered"""
+
+    def __init__(self, *, root_path: os.PathLike):
+        self.root_path = root_path if isinstance(root_path, raw_nodes.URI) else pathlib.Path(root_path).resolve()
+
+    def generic_visit(self, node):
+        if isinstance(node, raw_nodes.RawNode):
+            for field, expected_sha256 in iter_fields(node):
+                if expected_sha256 is missing:
+                    continue
+
+                if field == "sha256":
+                    source_name = "source"
+                    for sn in ["source", "uri"]:
+                        if hasattr(node, sn):
+                            source_name = sn
+                            break
+
+                elif field.endswith("_sha256"):
+                    source_name = field[: -len("_sha256")]
+                elif "sha256" in field:
+                    raise NotImplementedError(f"Don't know how to check integrity with {field}")
+                else:
+                    continue
+
+                if not hasattr(node, source_name):
+                    raise ValueError(
+                        f"Node {node} expected to have '{source_name}' field associated with '{expected_sha256}'"
+                    )
+
+                source_node = getattr(node, source_name)
+                source = get_resolved_source_path(source_node, root_path=self.root_path)
```

Comment:
not a super strong opinion at all, but I think readability here could be improved. `field`, `expected_sha256` are more like key, value, and only at after all the checking we're sure what we have. It would be great to have "correct" meaning from the start- so I'd probably factor out a generator that only returns fields that are `sha256`-like fields.